polygeist/tools/cgeist/Lib/CodeGenTypes.cc

```diff
@@ -1532,6 +1532,14 @@ mlir::Type CodeGenTypes::getMLIRType(clang::QualType qt, bool *implicitRef,
       return builder.getF80Type();
     if (T->isFP128Ty())
       return builder.getF128Type();
+    if (T->is16bitFPTy()) {
+      if (CGM.getTarget().shouldEmitFloat16WithExcessPrecision()) {
+        llvm::WithColor::warning()
```

Comment:
I think in addition to the warning we should generate a fp32 MLIR type here. But ideally the promotion should have been done already.

Reply:
Also what about complex types? I see this code in clang which makes me think one could have a complex floating point type with half-precision float element type: 

```
  QualType getPromotionType(QualType Ty) {
    if (CGF.getTarget().shouldEmitFloat16WithExcessPrecision()) {
      if (Ty->isAnyComplexType()) {
        QualType ElementType = Ty->castAs<ComplexType>()->getElementType();
        if (ElementType->isFloat16Type())
          return CGF.getContext().getComplexType(CGF.getContext().FloatTy);
      }
      if (Ty->isFloat16Type())
        return CGF.getContext().FloatTy;
    }
    return QualType();
  }
```

Reply:
> I think in addition to the warning we should generate a fp32 MLIR type here. But ideally the promotion should have been done already.

The drawback on doing that is that we need to test this as long as we add these conversions. I'm currently working on that.

Reply:
> Also what about complex types? I see this code in clang which makes me think one could have a complex floating point type with half-precision float element type:

This should be working fine, also with structs, vectors, etc., as they obtain MLIR types from this same function.

Reply:
So, if that's your concern:
```c
_Float16 _Complex z;
```
would generate the same warning as the non-complex case.